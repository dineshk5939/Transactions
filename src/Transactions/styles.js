import { withStyles, makeStyles } from '@material-ui/core/styles';
import { TableCell, TableRow } from '@material-ui/core';

const useStyles = makeStyles((theme) => ({
  root: {
    padding: 20,
    background: '#f5f5f5',
    height: '100vh',
  },
  grid: {
    margin: '20px 0',
  },
  tableContainer: {
    margin: '20px 0',
    boxShadow: '1px 2px 7px 2px rgba(0,0,0,0.2)',
  },
  table: {
    minWidth: 700,
  },
}));

const StyledTableCell = withStyles((theme) => ({
  head: {
    backgroundColor: theme.palette.common.black,
    color: theme.palette.common.white,
  },
  body: {
    fontSize: 14,
  },
}))(TableCell);

const StyledTableRow = withStyles((theme) => ({
  root: {
    '&:nth-of-type(odd)': {
      backgroundColor: theme.palette.action.hover,
    },
  },
}))(TableRow);

export { useStyles, StyledTableCell, StyledTableRow };
