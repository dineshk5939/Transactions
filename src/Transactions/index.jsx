import React, { useEffect, useState } from 'react';
import {
  Grid,
  Button,
  Typography,
  Select,
  MenuItem,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Paper,
} from '@material-ui/core';
import { useStyles, StyledTableRow, StyledTableCell } from './styles';
import axios from 'axios';
import map from 'lodash/map';

const baseUrl = 'http://localhost:5000/api';
const months = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];
const Transactions = () => {
  const classes = useStyles();
  const [transactions, setTransactions] = useState([]);
  const [users, setUsers] = useState([]);
  const [selectedUser, setSelectedUser] = useState('select');
  const [monthlyTotals, setMonthlyTotals] = useState({});
  const [totalRewardsPoints, setTotalRewardsPoints] = useState(0);

  const calculateRewardPoints = (transaction) => {
    let { amount } = transaction,
      rewardPoints = 0;

    if (amount > 50) {
      rewardPoints += amount < 100 ? amount - 50 : 50;
    }

    if (amount > 100) {
      rewardPoints += (amount - 100) * 2;
    }

    return {
      ...transaction,
      rewardPoints,
    };
  };

  const fetchUsers = () => {
    axios.get(`${baseUrl}/users`).then(({ data: { users } }) => {
      setUsers(users);
      setSelectedUser(users[0].id);
    });
  };

  const fetchTransactions = () => {
    axios
      .get(`${baseUrl}/transactions/${selectedUser}`)
      .then(({ data: { transactions } }) => {
        const monthlyTotals = {};
        let totalRewardsPoints = 0;
        const transactionsWithRewardPoints = map(transactions, (transaction) =>
          calculateRewardPoints(transaction),
        );
        map(transactionsWithRewardPoints, (transaction) => {
          const month = new Date(transaction.transactionDate).getMonth();
          monthlyTotals[month] = monthlyTotals[month] || 0;
          monthlyTotals[month] += transaction.rewardPoints;
          totalRewardsPoints += transaction.rewardPoints;
        });
        setTransactions(transactionsWithRewardPoints);
        setMonthlyTotals(monthlyTotals);
        setTotalRewardsPoints(totalRewardsPoints);
      });
  };

  useEffect(() => {
    fetchUsers();
  }, []);

  return (
    <Grid className={classes.root}>
      <Grid item sm={12} className={classes.grid}>
        <Select
          value={selectedUser}
          onChange={({ target: { value } }) => {
            setSelectedUser(value);
          }}
        >
          {map(users, ({ id, name }) => (
            <MenuItem key={id} value={id}>
              {name}
            </MenuItem>
          ))}
        </Select>
        <Button sm={6} color="primary" onClick={fetchTransactions}>
          {' '}
          Click to fetch the details of user{' '}
        </Button>
      </Grid>

      <Grid item sm={12} className={classes.grid}>
        <Typography>Total Reward Points: {totalRewardsPoints}</Typography>
      </Grid>

      <TableContainer component={Paper} className={classes.tableContainer}>
        <Table className={classes.table}>
          <TableHead>
            <TableRow>
              <TableCell>Name</TableCell>
              <TableCell align="right">Amount</TableCell>
              <TableCell align="right">Transaction Date</TableCell>
              <TableCell align="right">Reward Points</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {map(transactions, (transaction, i) => (
              <StyledTableRow key={`${transaction.id}_${i}`}>
                <StyledTableCell component="th" scope="row">
                  {' '}
                  {transaction.name}{' '}
                </StyledTableCell>
                <StyledTableCell align="right">
                  {transaction.amount}{' '}
                </StyledTableCell>
                <StyledTableCell align="right">
                  {transaction.transactionDate}{' '}
                </StyledTableCell>
                <StyledTableCell align="right">
                  {transaction.rewardPoints}{' '}
                </StyledTableCell>
              </StyledTableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>

      <TableContainer component={Paper} className={classes.tableContainer}>
        <Table className={classes.table}>
          <TableHead>
            <TableRow>
              <TableCell align="center">Month</TableCell>
              <TableCell align="center">Reward Points</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {map(monthlyTotals, (points, month) => (
              <StyledTableRow key={`${month}`}>
                <StyledTableCell align="center" component="th" scope="row">
                  {' '}
                  {months[month]}{' '}
                </StyledTableCell>
                <StyledTableCell align="center">{points} </StyledTableCell>
              </StyledTableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
    </Grid>
  );
};

export default Transactions;
