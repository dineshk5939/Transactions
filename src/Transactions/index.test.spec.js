import React from 'react';
import renderer from 'react-test-renderer';
import Transactions from './index';

it('renders correctly', () => {
  const tree = renderer.create(<Transactions />).toJSON();
  expect(tree).toMatchSnapshot();
});
