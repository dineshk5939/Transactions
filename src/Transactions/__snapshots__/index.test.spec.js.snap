// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiGrid-root makeStyles-root-1 MuiGrid-container"
>
  <div
    className="MuiGrid-root makeStyles-grid-2 MuiGrid-item MuiGrid-grid-sm-12"
  >
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        tabIndex={-1}
        value="select"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      sm={6}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
         Click to fetch the details of user 
      </span>
    </button>
  </div>
  <div
    className="MuiGrid-root makeStyles-grid-2 MuiGrid-item MuiGrid-grid-sm-12"
  >
    <p
      className="MuiTypography-root MuiTypography-body1"
    >
      Total Reward Points: 
      0
    </p>
  </div>
  <div
    className="MuiPaper-root MuiTableContainer-root makeStyles-tableContainer-3 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-4"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            Name
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Amount
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Transaction Date
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Reward Points
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
  <div
    className="MuiPaper-root MuiTableContainer-root makeStyles-tableContainer-3 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      className="MuiTable-root makeStyles-table-4"
      role={null}
    >
      <thead
        className="MuiTableHead-root"
        role={null}
      >
        <tr
          className="MuiTableRow-root MuiTableRow-head"
          role={null}
        >
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            Month
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter"
            scope="col"
          >
            Reward Points
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
        role={null}
      />
    </table>
  </div>
</div>
`;
